crates/astria-composer/src/searcher/bundler.rs

```diff
@@ -0,0 +1,135 @@
+use std::{
+    collections::VecDeque,
+    future::{
+        self,
+        Ready,
+    },
+    mem,
+};
+
+use astria_core::sequencer::v1alpha1::{
+    transaction::{
+        action::SequenceAction,
+        Action,
+    },
+    ROLLUP_ID_LEN,
+};
+use tracing::debug;
+
+#[derive(Debug, thiserror::Error)]
+enum SizedBundleError {
+    #[error("bundle does not have enough space left for the given sequence action")]
+    NotEnoughSpace(SequenceAction),
+}
+
+/// A bundle sequence actions to be submitted to the sequencer. Maintains the total size of the
+/// bytes pushed to it and enforces a max size in bytes passed in the constructor. If an incoming
+/// `seq_action` won't fit in the buffer it is flushed and a new bundle is started.
+pub(super) struct SizedBundle {
+    /// The buffer of actions
+    pub(super) buffer: Vec<Action>,
+    /// The current size of the bundle in bytes. This is equal to the sum of the size of the
+    /// `seq_action`s + `ROLLUP_ID_LEN` for each.
+    pub(super) curr_size: usize,
+    /// The max bundle size in bytes to enforce.
+    max_size: usize,
+}
+
+impl SizedBundle {
+    /// Create a new empty bundle with the given max size.
+    pub(super) fn new(max_size: usize) -> Self {
+        Self {
+            buffer: vec![],
+            curr_size: 0,
+            max_size,
+        }
+    }
+
+    /// Buffer `seq_action` into the bundle. If the bundle won't fit `seq_action`, flush `buffer`,
+    /// returning it, and start building up a new buffer using `seq_action`.
+    fn push(&mut self, seq_action: SequenceAction) -> Result<(), SizedBundleError> {
+        let seq_action_size = seq_action.data.len() + ROLLUP_ID_LEN;
```

Comment:
two points:
1. i'm seeing this many places, should we make a function for it?
2. this doesn't take into account encoding size, should we have some amount of buffer for that?

Reply:
noted in comments